src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -874,7 +874,9 @@ void DiskObjectStorageTransaction::writeFileUsingBlobWritingFunction(
     /// Create metadata (see create_metadata_callback in DiskObjectStorageTransaction::writeFile()).
     if (mode == WriteMode::Rewrite)
     {
-        if (!object_storage.isWriteOnce() && metadata_storage.exists(path))
+        /// Otherwise we will produce lost blobs which nobody points to
+        /// WriteOnce storages are not affected by the issue
+        if (!object_storage.isPlain() && metadata_storage.exists(path))
```

Comment:
This is the first part of the fix. `object_storage.isWriteOnce()` returns false for `s3_plain_rewritable` disks, so it could actually remove restored files immediately after copying them to a destination disk.

Checking `object_storage.isPlain()` instead is more correct here and it also corresponds to [DiskObjectStorageTransaction::writeFile()](https://github.com/ClickHouse/ClickHouse/blob/846922aeae590d4c441404f1874a4ec4fad22022/src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp#L805).